### PR TITLE
Issue multiple warp sync requests in grandpa warp sync if needed

### PR DESCRIPTION
**Pull request by expenses**
... And other warp sync changes

### CODE REVIEW

**Comment by tomaka** [comment]
Instead of a hardcoded limit of 100, is it not possible to continue issuing requests until it comes back empty?

**Comment by expenses** [comment]
> Instead of a hardcoded limit of 100, is it not possible to continue issuing requests until it comes back empty?

I'm not sure, it might be that the minimum number of requests returned is 1. I would much rather send the fragments along with a boolean that indicates whether more requests need to be made or not though.

**Comment by tomaka** [comment]
The boolean would indeed remove a round trip time.

> I'm not sure, it might be that the minimum number of requests returned is 1.

Note that we must be resilient to that kind of thing if the Substrate node misbehaves (intentionally or not). Smoldot shouldn't be trapped in an infinite loop, whatever what is received is.

**Comment by expenses** [comment]
>  I'm not sure, it might be that the minimum number of requests returned is 1.

I just confirmed this by setting `let final_set_of_fragments = response_fragments.len() == 1;`.

> Note that we must be resilient to that kind of thing if the Substrate node misbehaves (intentionally or not). Smoldot shouldn't be trapped in an infinite loop, whatever what is received is.

I think that having the check above would eliminate most problems.

**Comment by tomaka** [comment]
Is it not possible for the target Substrate node to randomly generate one Grandpa warp sync fragment (with a different random hash every time) and block smoldot forever?

**Comment by expenses** [comment]
> Is it not possible for the target Substrate node to randomly generate one Grandpa warp sync fragment (with a different random hash every time) and block smoldot forever?

I don't think verification would succeed in this case. And as `final_set_of_fragments` would be true, we would only attempt to validate once. What _might_ block smoldot is if a node sent the same (valid) fragment 100 times. I'm not sure what happens with verification then.

**Comment by tomaka** [comment]
Right, sorry. I somehow had in my mind that we were downloading everything first, and then only start the verification.

File: src/chain/sync/grandpa_warp_sync.rs [hunk 1]
```diff
@@ -155,6 +157,11 @@ impl<TSrc> StorageGet<TSrc> {
         &self.state.warp_sync_source
     }
 
+    /// Returns the header that we're warp syncing up to.
+    pub fn warp_sync_header(&self) -> Header {
+        self.state.header.clone()
+    }
```
Comment (tomaka) [inline src/chain/sync/grandpa_warp_sync.rs:163]: Let's please not just copy things all the time.

```suggestion
    pub fn warp_sync_header(&self) -> HeaderRef {
        (&self.state.header).into()
    }
```

File: src/chain/sync/grandpa_warp_sync.rs [hunk 2]
```diff
@@ -194,6 +201,11 @@ impl<TSrc> NextKey<TSrc> {
...
```
Comment (tomaka) [inline src/chain/sync/grandpa_warp_sync.rs:207]: Same

File: src/chain/sync/grandpa_warp_sync.rs [hunk 3]
```diff
@@ -340,6 +402,11 @@ pub struct VirtualMachineParamsGet<TSrc> {
 }
 
 impl<TSrc> VirtualMachineParamsGet<TSrc> {
+    /// Returns the header that we're warp syncing up to.
+    pub fn warp_sync_header(&self) -> Header {
+        self.state.header.clone()
+    }
```
Comment (tomaka) [inline src/chain/sync/grandpa_warp_sync.rs:408]: Ditto

File: src/finality/grandpa/warp_sync.rs [hunk 4]
```diff
@@ -32,10 +32,10 @@ pub struct Verifier {
 
 impl Verifier {
     pub fn new(
-        start_chain_information: &ChainInformation,
+        start_chain_information_finality: &ChainInformationFinality,
```
Comment (tomaka) [inline src/finality/grandpa/warp_sync.rs:35]:
```suggestion
        start_chain_information_finality: ChainInformationFinalityRef,
```

**Review by tomaka — CHANGES_REQUESTED** [review]
Good after removing the extra copies

**Review by tomaka — APPROVED** [review]

**Pull request merged**